content/post/2020-05-07-solaris/index.Rmd

```diff
@@ -0,0 +1,155 @@
+   and choose one of our Solaris builders.
+
+2. Bookmark this page, in case you get an email from CRAN about your
+   package failing on Solaris.
+
+## Oracle Solaris
+
+[Oracle Solaris](https://en.wikipedia.org/wiki/Solaris_\(operating_system\))
+is a non-free Unix operating system. CRAN regularly tests R packages on
+Solaris 10. This is a post about checking that your R package on
+Solaris, using R-hub, and how you can create your own Solaris virtual
+machine to debug your R code.
+
+## OpenCSW
+
+[OpenCSW](https://www.opencsw.org/about/) is a repository of open source
+software packages. It uses a package manager tool called `pkgutil`.
+`pkgutil` is similar to `apt`, `yum`, etc.
+
+OpenCSW greatly varies in terms of how current its various packages are.
+By default OpenCSW uses its "testing" distribution, which has a rolling
+release model. The most recently added packages are in the "unstable"
+distribution. OpenCSW unstable has now a new R 4.0.0 build, called
+`r_base`. It will move to the "testing" distibution in a couple of days.
+OpenCSW `r_base` 4.0.0 is one of the R builds used on R-hub's Solaris
+builder.
+
+R-hub also has a repository (catalog) of OpenCSW packages,
+which we use for system requirements that are not (yet) updated in the
+official OpenCSW catalog. This catalog is at
+https://files.r-hub.io/opencsw/
+
+## CRAN's Solaris builder
+
+CRAN uses two sets of compilers for their R package checks.
+[Oracle Developer Studio](https://www.oracle.com/tools/developerstudio/)
+(ODS) is a commercial product that supports Oracle Solaris 10. In addition,
+[OpenCSW](https://www.opencsw.org/about/) packages GCC, version 5.5.0
+currently.
+
+CRAN compiles R packages with ODS by default. If ODS is not able to
+compile a package, they use GCC instead. GCC should be able to compile
+most (all?) CRAN R packages. Notably, Rcpp and all packages linking to it
+are compiled with GCC.
+
+CRAN's configuration of the two compilers is here:
+<https://www.stats.ox.ac.uk/pub/bdr/Rconfig/r-patched-solaris-x86>
+
+## R-hub's Solaris builder
+
+R-hub's setup is similar. Our Solaris virtual machine has three R
+versions:
+* 32-bit GCC build at `/opt/csw/bin/R`, from the OpenCSW `r_base` package.
+  The configuration of this build is in OpenCSW, see here:
+  <https://buildfarm.opencsw.org/source/xref/opencsw/csw/mgar/pkg/r/trunk/Makefile>
+* 64-bit GCC build at `/opt/csw/bin/amd64/R`, from the same OpenCSW
+  `r_base` package.
+* 32-bit ODS build, using the `config.site` file at
+  <https://github.com/r-hub/solarischeck/blob/master/config.site.cc>.
+
+If you call `rhub::platforms()` you'll currently see two Solaris builders:
+* `solaris-x86-patched` uses the 32-bit GCC build of R.
+* `solaris-x86-patched-ods` uses the 32-bit GCC build to install all
+  dependencies of the package, but then runs `R CMD check` with the ODS
+  build. This maximizes the chance of a meaningful `R CMD check` run.
+
```

Comment:
What about https://r-hub.github.io/rhub/reference/check.html?

Reply:
What about it?

Reply:
I mean the solaris shortcut, should it be mentioned?

Reply:
Oh, right, sure. These client side shortcuts are not so great, as it turns out, because now it only has the first Solaris builder, but not the second one. Also, it is actually R-release now, but I could not change the name, because of the shortcut....